### PR TITLE
Add peeling to Single and DoubleTilingExpert

### PR DESCRIPTION
**Pull request by dcaballe**
This PR relies on the peeling API introduced in the Codegen
Strategy infrastructure (https://reviews.llvm.org/D126842)
to enable peeling in the single and double tiling experts.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ static llvm::cl::opt<bool> useLinalgTransformInterp(
 static llvm::cl::opt<bool> disableMatmulPadPipeline(
     "iree-codegen-disable-matmul-pad-pipeline",
     llvm::cl::desc("disable padding options in Matmul codegen"),
-    llvm::cl::init(false));
+    llvm::cl::init(true));
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:76]: We need to discuss how to integrate this with padding

File: compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp [hunk 2]
```diff
@@ -305,6 +306,45 @@ void LinalgFusePass::runOnOperation() {
   }
 }
 
+/// Return true if 'loop' is a tiled loop that distributes work among threads.
+/// We currently check whether the loop control depends on any of the interface
+/// workgroup operations. This may lead to some false positives that may prevent
+/// peeling of loops that we could peel. We shouldn't need this check when we
+/// move to the Transform dialect.
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:313]: The main problem with the selection of loops to peel is that we may tile loops for vectorization that are later removed if they have a single iteration. (IMO, we should be doing this kind of loop optimizations at the end and not in the middle of the strategy but that's a discussion for another day 😄). The current implementation seems to be enough and fall on the conservative side but I'm open to other suggestions.

**Comment by dcaballe** [comment]
Some tests should fail since I'm disabling padding. Let's see how bad the performance is.

File: compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:313]: Yeah, that is a issue with throwing everything into the strategy.

But you should be able to use this (does a similar traversal) https://github.com/google/iree/blob/e2876535b5208f012b97b92c8a058ae9803d48ac/compiler/src/iree/compiler/Codegen/Utils/Utils.h#L160

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 3]
```diff
@@ -342,6 +342,7 @@ void addDoubleTilingExpertPassPipeline(OpPassManager &passManager,
   // Add the sandbox single tiling expert to tile and vectorize.
   {
     LinalgSingleTilingExpertPassOptions options;
+    options.peel = true;
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:345]: If we add a new pipeline, we can control this option with by passing the value (true/false) from the caller. For the pipeline with peeling we use `true`.
I dont think peeling and padding go together right (at least for now)

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 3 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:345]: Yes. However, I think I've lost track of the current state of vectorization. Padding is only enabled for matmul but we added a few fixes to "always vectorized" (setting tile sizes to one) and other changes. Could you please give me a quick update? I understand that we should enable peeling for any non-matmul op, right?

> I dont think peeling and padding go together right (at least for now)

No, they wouldn't work together. I run this to get an overall performance picture an see how things evolve with the upcoming changes.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:345]: Padding is only enabled for matmul. For other static cases, we set the tiling size to factors of dim sizes. If they are dynamic cases, the tiling sizes are set to one.

I think we can add a new pipeline for peeling. After more exploration, we can configure the kernel to use which pipeline in KernelConfig.cpp.

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:345]: Ok, let me move the state of this PR to draft to run some experiments before creating the new pipeline.

**Comment by dcaballe** [comment]
That should be all in terms of having the basic infra to enable peeling. No peeling is enable by default right now. Looking into enabling it for dynamic shapes and element-wise ops for now. I'll send separate patches to see if there is any performance impact.

A couple of sort of unrelated comments:

1. Now that we have both padding and peeling implementations in places, we should consider unifying the two `addDoubleTiling*Expert` functions to reduce the maintainability burden. Otherwise, we will end up building and maintaining two separate pipelines. Padding and peeling shouldn't be a reason for that.

2. We should also consider taking some steps to simplify the the logic used to set up the tile sizes for vectorization. Currently it's a bit scattered over several places and it gets difficult to actually know what the final state of the tile sizes will be. I think we should try to refactor this logic to one place and then make the tile sizes container constant so that it can't be further modified. Probably not something that we can do in one shot but hopefully we can find opportunities to move towards that direction.

**Comment by hanhanW** [comment]
I think moving to transform dialect can address the latter point.

> We should also consider taking some steps to simplify the the logic used to set up the tile sizes for vectorization. Currently it's a bit scattered over several places and it gets difficult to actually know what the final state of the tile sizes will be.

There are parallel tiling sizes and reduction tiling sizes today. In the beginning, we didn't want to modify CodegenStrategy while it's not stable yet. It's hard to separate them when adding the passes to the pass manger. Today, we should be able to merge them into `WorkgroupTilingSizes` which is a list of ints, and keep the logic of extracting tiling sizes in scheduler (which is CodegenStrategy today). Since we'll migrate to use transform dialect, we can simplify the logic when creating transform ops.

> I think we should try to refactor this logic to one place and then make the tile sizes container constant so that it can't be further modified. Probably not something that we can do in one shot but hopefully we can find opportunities to move towards that direction.

I expect that the tiling sizes will be frozen if we move to use transform dialect. We'll build the scheduler, and they will be embedded in attributes of transform ops. Then we'll kick start the transforms right after scheduling. At this stage, we're not able to modify tiling sizes anymore.

**Review by hanhanW — APPROVED** [review]
Overall looks good to me!

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**